### PR TITLE
VCL snippet upload

### PR DESCRIPTION
**Pull request by tomasinchoo**

### CODE REVIEW

File: src/Api.php [hunk 1]
```diff
@@ -133,7 +131,7 @@ public function validateApiKey() {
    * Gets a list of services for the current customer.
    */
   public function getServices() {
-    $response = $this->query('service');
+    $response = $this->query('/service');
```
Comment (greylabel) [inline src/Api.php:143]: Good catch.

File: src/Api.php [hunk 2]
```diff
@@ -304,6 +302,60 @@ protected function query($uri, array $data = [], $method = 'GET', array $headers
...
+   * @param array $headers
+   *   (optional) An array of headers to send with the request.
+   *
+   * @return \Psr\Http\Message\ResponseInterface
+   *   Response.
+   *
+   * @throws \GuzzleHttp\Exception\RequestException
+   *   RequestException.
+   */
+  protected function VQuery($uri, array $data = [], $method = 'GET', array $headers = []) {
```
Comment (greylabel) [inline src/Api.php:332]: Is this redundant with the `query` method? Should we refactor/combing the two?

File: src/Form/FastlySettingsForm.php [hunk 3]
```diff
@@ -41,11 +49,14 @@ class FastlySettingsForm extends ConfigFormBase {
    *   Fastly API for Drupal.
    * @param \Drupal\fastly\State $state
    *   Fastly state service for Drupal.
+   * @param \Drupal\fastly\VclHandler
+   *   Vcl handler
    */
-  public function __construct(ConfigFactoryInterface $config_factory, Api $api, State $state) {
+  public function __construct(ConfigFactoryInterface $config_factory, Api $api, State $state, VclHandler $vclHandler) {
```
Comment (greylabel) [inline src/Form/FastlySettingsForm.php:55]: 👍 For dependency injection being used correctly.

File: src/VclHandler.php [hunk 4]
```diff
@@ -0,0 +1,739 @@
...
+   * Sets data to be processed, sets Credentials
+   * Vcl_Handler constructor.
+   *
+   * @param ConfigFactoryInterface $config_factory
+   * @param $host
+   * @param Api $api
+   */
+  public function __construct(ConfigFactoryInterface $config_factory, $host, Api $api) {
+    $vcl_dir = drupal_get_path('module', 'fastly'). '/vcl_snippets';
+    $data = array(
```
Comment (greylabel) [inline src/VclHandler.php]: Drupal coding standards use short array syntax `[]`.

File: src/VclHandler.php [hunk 5]
```diff
@@ -0,0 +1,739 @@
...
+         // send_web_hook($message);
+        }
+      } elseif ($pass && !$activate) {
+        $message = 'VCL updated, but not activated.';
+      }
+
+    } catch (Exception $e) {
+      $this->addError('Some of the API requests failed, enable debugging and check logs for more information.');
+      $message = 'VCL update failed : ' . $e->getMessage();
+      error_log($message);// Force log this, possibly no response object
```
Comment (greylabel) [inline src/VclHandler.php]: Should you be using a injected Drupal logger channel here.

File: fastly.services.yml [hunk 6]
```diff
@@ -28,3 +28,6 @@ services:
   logger.channel.fastly:
     parent: logger.channel_base
     arguments: ['fastly']
+  fastly.vclhandler:
+    class: Drupal\fastly\VclHandler
+    arguments: ['@config.factory', '%fastly.host%', '@fastly.api']
```
Comment (greylabel) [inline fastly.services.yml]: Might want to pass `'@logger.channel.fastly` to you can use it in your new service.

File: src/VclHandler.php [hunk 7]
```diff
@@ -0,0 +1,739 @@
...
+   * @param $url
+   * @param array $headers
+   * @param array $data
+   * @return \Psr\Http\Message\ResponseInterface
+   */
+  public function vclPutWrapper($url, $headers = array(), $data = array()) {
+    return $this->vclRequestWrapper($url, $headers, $data, "PUT");
+  }
+
+  public function log($level, $message) {
```
Comment (greylabel) [inline src/VclHandler.php]: Using `\Drupal::` in a service in not best practice -- use injected logger.

File: src/VclHandler.php [hunk 8]
```diff
@@ -0,0 +1,739 @@
...
+      if ($response->getStatusCode() == "200") {
+        $status = true;
+        $response_body = json_decode($response->getBody());
+        $service_name = $response_body->name;
+        $message = 'Connection Successful on service *' . $service_name . "*";
+      } else {
+        $status = false;
+        $response_body = json_decode($response->getBody());
+        $service_name = $response_body->name;
+        $message = 'Connection not Successful on service *' . $service_name . "*" . " status : " . $response->getStatusCode();
```
Comment (greylabel) [inline src/VclHandler.php]: Look at using placeholder replacement in the log message:

https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Component%21Render%21FormattableMarkup.php/function/FormattableMarkup%3A%3AplaceholderFormat/8.3.x

https://drupalize.me/blog/201510/how-log-messages-drupal-8

File: src/VclHandler.php [hunk 9]
```diff
@@ -0,0 +1,739 @@
...
+
+  /**
+   * Function for testing Fastly API connection
+   *
+   * @param $hostname
+   * @param $service_id
+   * @param $api_key
+   * @return array
+   */
+  public function testFastlyApiConnection($hostname, $service_id, $api_key) {
```
Comment (greylabel) [inline src/VclHandler.php]: Thoughts on this being a method in the main API class?

**Review by greylabel — COMMENTED** [review]
I did a quick read through. I will try to find some more time to test this completely. Made a few notes mostly around style and formatting.

File: src/Api.php (hunk 2 above)
Comment (tomasinchoo) [reply src/Api.php:332]: Yes, I believe this needs some additional adjusting. Note that I'm fairly new to Drupal 8 (I use to work with it very long time ago so I'm still catching up).
In this module I was actually trying to port vcl functionality from WP (https://github.com/fastly/WordPress-Plugin/blob/master/src/classes/vcl-handler.php) and this ended up little bit more tricky than I initially thought. Being new to fastly, this looked like good idea at the time.
Anyway, I'm little more into Drupal and Fastly now and I'll hurry up with changes.
Thanks for very helpful comments.

Comment (greylabel) [reply src/Api.php:332]: Sounds good and thanks for working on these issues. I am happy to chat through ideas or help with Drupal 8 best practices as needed.

**Comment by tomasinchoo** [comment]
@greylabel @vvuksan 
I made most of changes requested in last pull request, however, I'm still not 100% sure if it is good idea to refactor VQuery and query at this point.
I am mostly concerned with this piece of code:
`switch (strtoupper($method)) {
        case 'GET':
          $uri = ltrim($uri, "/");
          return $this->httpClient->request($method, $this->host . $uri, $data);

        case 'POST':
          return $this->httpClient->post($this->host . $uri, $data);

        case 'PURGE':
          return $this->httpClient->request($method, $uri, $data);
        default:
          throw new \Exception('Method :method is not valid for Fastly service.', [
            ':method' => $method,
          ]);
      }`
Thing is that for example I'm not aware that guzzle has method "post" that is used here. Also, GET and PURGE methods recieve different parameters (one require URL, other require URI).
To refactor this properly, whole module should be investigated and fixed to introduce some consistency in how this method is used. I already had some issues with how URL/URI should be constructed in the module (i.e. $uri = ltrim($uri, "/");) etc.
This could bring some additional complexity and it wouldn't be really clear which part of merge makes originaly requested feature.
My suggestion is to stick with how it is now, and introduce new task for refactoring the module.
Let me know what you think?

**Comment by greylabel** [comment]
> My suggestion is to stick with how it is now, and introduce new task for refactoring the module.
> Let me know what you think?

I am not opposed to that approach, but definitely think we should follow through with the refactor if we merge this and make sure to should consider in the refactor the `fastlypurger` module that is part of the project.

**Comment by greylabel** [comment]
@tomasinchoo There are merge conflicts now that we have merged your other two recent PRs.

**Pull request merged**